components/suggest/src/yelp.rs

```diff
@@ -0,0 +1,322 @@
+        scope.err_if_interrupted()?;
+        conn.execute(
+            "INSERT INTO yelp_location_signs(keyword, need_location) VALUES(:keyword, :need_location)",
+            named_params! {
+                ":keyword": sign.keyword,
+                ":need_location": sign.need_location,
+            },
+        )?;
+    }
+
+    *URL_TEMPLATE.lock().unwrap() = dataset.template.to_string();
+
+    Ok(())
+}
+
+/// Fetch Yelp suggestion from given user's query.
+pub(crate) fn fetch_suggestion(
+    conn: &Connection,
+    query: &SuggestionQuery,
+) -> Result<Option<Suggestion>> {
+    if !query.providers.contains(&SuggestionProvider::Yelp) {
+        return Ok(None);
+    }
+
+    if query.keyword.len() > MAX_QUERY_LENGTH {
+        return Ok(None);
+    }
+
+    if !&query.keyword.contains(" ") {
+        // Try to create by one word.
+        return to_suggestion(conn, query, &query.keyword, None, None, None, None, false);
+    }
+
+    // Find the location sign and the location.
+    let (query_without_location, location_sign, location, need_location) =
+        find_location(conn, &query.keyword)?;
+
+    if let (Some(_), false) = (&location, need_location) {
+        // The location sign does not need the specific location, but user is setting something.
+        return Ok(None);
+    }
+
+    if query_without_location.is_empty() {
+        // No remained query.
+        return Ok(None);
+    }
+
+    // Find the modifiers.
+    let (subject_candidate, pre_modifier, post_modifier) =
+        find_modifiers(conn, &query_without_location)?;
+
+    to_suggestion(
+        conn,
+        query,
+        &subject_candidate,
+        pre_modifier,
+        post_modifier,
+        location_sign,
+        location,
+        need_location,
+    )
+}
+
+fn to_suggestion(
+    conn: &Connection,
+    query: &SuggestionQuery,
+    subject: &str,
+    pre_modifier: Option<String>,
+    post_modifier: Option<String>,
+    location_sign: Option<String>,
+    location: Option<String>,
+    need_location: bool,
+) -> Result<Option<Suggestion>> {
+    if !is_subject(conn, subject)? {
+        // No subject.
+        return Ok(None);
+    }
+
+    // Build URL.
+    let mut description = String::from(subject);
+
+    if let Some(modifier) = &pre_modifier {
+        description.insert(0, ' ');
+        description.insert_str(0, &modifier);
+    }
+
+    if let Some(modifier) = &post_modifier {
+        description.push(' ');
+        description.push_str(&modifier);
+    }
+
+    if let (Some(sign), false) = (&location_sign, need_location) {
+        // This location sign such the 'near by' needs to add as a description parameter.
+        description.push(' ');
+        description.push_str(&sign);
+    }
+
+    let mut url = URL_TEMPLATE.lock().unwrap().clone();
+    url = url.replace(
+        "{description}",
+        &form_urlencoded::byte_serialize(description.as_bytes())
+            .collect::<Vec<_>>()
+            .concat(),
```

Comment:
There's a [`Serializer` struct](https://docs.rs/form_urlencoded/latest/form_urlencoded/struct.Serializer.html#method.finish) that might be a bit more ergonomic to use than converting back and forth to bytes.